File: Crossplay Ball game/JavaClient/src/GameClient.java

import java.io.*;
import java.net.Socket;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.text.SimpleDateFormat;
import java.util.Arrays;
import java.util.Calendar;

public class GameClient extends BallApp {

    private Socket socket;
    private String server;
    private BufferedReader sInput;
    private PrintWriter sOutput;
    private int port, id, ballHolder;
    boolean ball;

    GameClient(String server, int port) {
        this.server = server;
        this.port = port;
        id = 0;
    }

    boolean start() {
        try {
            ball = false;
            connected = false;

            socket = new Socket(server, port);
            sInput = new BufferedReader(new InputStreamReader(socket.getInputStream()));
            sOutput = new PrintWriter(new OutputStreamWriter(socket.getOutputStream()));
        } catch (Exception ec) {
            feed.setText("Error connecting to server:" + ec);
            return false;
        }
        feed.setText("\nConnection accepted " + socket.getInetAddress() + " : " + socket.getPort() + "\n");

        // creates the Thread to listen from the server
        new ListenFromServer().start();
        connected = true;
        if (restart)
            id = backUp[0];
        sendMessage("ID " + id);
        return true;
    }

    void display(String msg) {
        try {
            feed.getDocument().insertString(0, "\n" + msg + "\n", null);
            feed.setCaretPosition(0);
        } catch (Exception ignored) {
        }
    }

    void sendMessage(String i) {
        sOutput.println(i);
        sOutput.flush();

        if (i.contains("BALL")) {
            ball = false;
            sendMessage("LIST " + 0);
        }

    }

    private void save() {
        backUp = new int[2];
        backUp[0] = id;
        backUp[1] = ball ? 1 : 0;
    }

    void disconnect() {
        try {
            if (sOutput != null)
                sOutput.close();
            if (socket != null)
                socket.close();

            users.setText("Active Users: \n(offline)");
            feed.setText("You have left the server");
        } catch (Exception ignored) {
        }
    }

    class ListenFromServer extends Thread {

        public synchronized void run() {

            while (true) {
                try {
                    String n = sInput.readLine();
                    if (n != null) {
                        String date = new SimpleDateFormat("HH:mm:ss").format(Calendar.getInstance().getTime());
                        String type = n.replaceAll("\\D+", "");
                        int user = Integer.parseInt(type);

                        if (n.contains("ID")) {
                            id = user;
                            frame.setTitle("Ball Game: Your id is " + id);
                            frame.setSize(650, 480);
                            display("> Your ID is : " + id);
                            info.setText("Welcome to the server:" +
                                    "\nIf you have the ball, enter an Id from the list & click Send" +
                                    "\nIf not, wait for it to be sent to you" +
                                    "\n" +
                                    "\nClick Refresh to update the active user list" +
                                    "\nClick Leave or close the app to leave the server");
                        } else if (n.contains("LIST")) {
                            if (user == 0)
                                users.setText("Active Users: \n" + date);
                            else {
                                String userLine = user + "";
                                if (user == id)
                                    userLine += " (You)";
                                if (user == ballHolder)
                                    userLine += " (Ball)";
                                users.append("\n> " + userLine);
                            }
                        } else if (n.contains("BALL")) {
                            ball = true;
                            ballHolder = id;
                            if (user != id) {
                                String userName = user + "";
                                if (user == 0)
                                    userName = "Server";
                                display("> " + date + ": Ball Received from " + userName);
                            } else
                                display("> " + date + ": Ball was returned");
                            display("> " + date + ": You now have the ball");
                            sendMessage("LIST " + 0);
                        } else if (n.contains("JOINED")) {
                            if (user != id)
                                display("> " + date + ": " + user + " has joined the game");
                            sendMessage("LIST " + 0);
                        } else if (n.contains("LEFT")) {
                            display("> " + date + ": " + user + " has left the game");
                            sendMessage("LIST " + 0);
                        } else if (n.contains("WHO")) {
                            ballHolder = user;
                            display("> " + date + ": " + ballHolder + " now has the ball");
                            sendMessage("LIST " + 0);
                        } else if (n.contains("RETURN")) {
                            ball = false;
                            ballHolder = user;
                            sendMessage("LIST " + 0);
                        }
                    }

                } catch (IOException e) {
                    if (connected && !restart) {
                        save();
                        join.doClick();
                        connected = false;
                        restart = true;
                        start.doClick();
                        display("Server has restarted & Game has continued");
                        display("> Your ID is still : " + id);
                        frame.setTitle("Ball Game: Your id is " + id);
                        info.setText("Welcome to the new server:" +
                                "\nIf you have the ball, enter an Id from the list & click Send" +
                                "\nIf not, wait for it to be sent to you" +
                                "\n" +
                                "\nClick Refresh to update the active user list" +
                                "\nClick Leave or close the app to leave the server");
                        frame.repaint();
                    }
                    break;
                }
            }
        }
    }
}

File: Crossplay Ball game/GameClient/GameClient/GameClient.cs
using System;
using System.Diagnostics;
using System.IO;
using System.Net.Sockets;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading;
using System.Web;

namespace GameClient
{
    public class GameClient : BallApp
    {
        public string server;
        public int port;
        public TcpClient socket;
        public NetworkStream stream;
        public static StreamReader sInput { get; private set; }
        public static StreamWriter sOutput { get; set; }
        public static BallApp app;
        public bool restart = false;
        public static bool ball;
        public static int id, ballHolder;

        public GameClient(string server, int port, BallApp ba)
        {
            this.server = server;
            this.port = port;
            app = ba;
            id = 0;
        }


        public bool Start()
        {
            
            {
                try
                {
                    ball = false;
                    socket = new TcpClient(server, port);
                    stream = socket.GetStream();
                    sInput = new StreamReader(stream);
                    sOutput = new StreamWriter(stream);
                }

                catch (Exception ec)
                {
                    DisplayFeed("Error connecting to server:" + ec);
                    return false;
                }
                DisplayNewFeed("Connection accepted " + socket.ToString());

                new ListenFromServer(this).Start();
                try
                {
                    if (restart)
                        id = backUp[0];
                }
                catch (Exception)
                {
                    return false;
                }
                SendMessage("ID " + id);
                return true;
            }
        }

        public static void DisplayFeed(string msg)
        {
            app.AppendFeed(msg);
        }


        public static void DisplayList(string msg)
        {
            app.AppendList(msg);
        }


        public static void DisplayNewFeed(string msg)
        {
            app.ReplaceFeed(msg);
        }


        public static void DisplayNewList(string msg)
        {
            app.ReplaceList(msg);
        }

        public static void SendMessage(string msg)
        {
            try
            {
                sOutput.WriteLine(msg);
                sOutput.Flush();
            }
            catch (Exception)
            { }

            if (msg.Contains("BALL"))
            {
                ball = false;
                SendMessage("LIST " + 0);
            }

        }


        public void Disconnect()
        {
            try
            {
                sOutput.Close();
                sInput.Close();
                socket.Close();
                Close();
            }
            catch (Exception) { }


        }

        public new void Leave()
        {
            Disconnect();
            app.LeaveSafe("Join");
            app.InputText("Localhost");
            app.LabelText("Enter Server's Address:");
            app.ReplaceFeed("> " + DateTime.Now.ToShortTimeString() + ": You have left the server");
            app.ReplaceList("Active Users: ");
            app.AppendList("(Offline)");
            app.TitleText("Ball Game");
            restart = false;
            connected = false;
            id = 0;
        }


        public void Save()
        {
            backUp = new int[2];
            backUp[0] = id;
            backUp[1] = ball ? 1 : 0;
        }

        public void SetTitle(string i)
        {
            app.TitleText(i);
        }

        


        abstract class CThread
        {
            private Thread _thread;

            protected CThread()
            {
                _thread = new Thread(new ThreadStart(RunThread));
            }

            // Thread methods / properties
            public void Start() => _thread.Start();
            public void Join() => _thread.Join();
            public bool IsAlive => _thread.IsAlive;

            // Override in base class
            public abstract void RunThread();
        }



        class ListenFromServer : CThread
        {
            GameClient gc;

            public ListenFromServer(GameClient gc)
            {
                this.gc = gc;
            }

            public override void RunThread()
            {
                while (true)
                {

                    try
                    {
                        string k = sInput.ReadLine();
                        string n = HttpUtility.UrlDecode(k, Encoding.UTF8);

                        string type = Regex.Match(n, @"\d+").Value;
                        int user = int.Parse(type);
                        string date = DateTime.Now.ToShortTimeString();

                        if (n.Contains("BALL"))
                        {
                            ball = true;
                            ballHolder = id;
                            if (user != id)
                                DisplayFeed("> " + date + ": Ball Received from " + user);
                            else DisplayFeed("> " + date + ": Ball was returned");
                            DisplayFeed("> " + date + ": You now have the ball");
                            SendMessage("LIST " + 0);
                        }
                        else if (n.Contains("LIST"))
                        {
                            if (user == 0)
                                DisplayNewList("Active Users: " + date);
                            else
                            {
                                string userLine = user + "";
                                if (user == id)
                                    userLine += " (You)";
                                if (user == ballHolder)
                                    userLine += " (BALL)";
                                DisplayList("> " + userLine);
                            }
                        }
                        else if (n.Contains("ID"))
                        {
                            id = user;
                            DisplayFeed("> Your ID is : " + id);
                            gc.SetTitle("Ball Game: Your ID is " + id);
                        }
                        else if (n.Contains("JOINED"))
                        {
                            if (user != id)
                                DisplayFeed("> " + date + ": " + user + " has joined the game");
                            SendMessage("LIST " + 0);
                        }
                        else if (n.Contains("LEFT"))
                        {
                            DisplayFeed("> " + date + ": " + user + " has left the game");
                            SendMessage("LIST " + 0);
                        }
                        else if (n.Contains("WHO"))
                        {
                            ballHolder = user;
                            DisplayFeed("> " + date + ": " + ballHolder + " now has the ball");
                            SendMessage("LIST " + 0);
                        }
                        else if (n.Contains("RETURN"))
                        {
                            ball = false;
                            ballHolder = user;
                            SendMessage("LIST " + 0);
                        }
                    }
                    catch (IOException)
                    {
                        if (connected) //If server restart
                        {
                            gc.Save();
                            gc.Disconnect();
                            
                            Process.Start(@"GameServer.exe");
                            gc = new GameClient("Localhost", 1500, app)
                            {
                                restart = true
                            };
                            if (gc.Start())
                            {
                                DisplayFeed("Server has restarted & Game continued");
                                DisplayFeed("> You id is still : " + backUp[0]);
                                if (backUp[1] == 1)
                                    SendMessage("RETURN " + backUp[0]);
                                gc.restart = false;
                            }
                            else
                            {
                                gc.Leave();
                            }
                        }
                        break;
                    }
                    catch (Exception)
                    {
                        break;
                    }

                }

            }
        }

    }
}

